ENH Improve initialization and learning rate in t-SNE

dkobak opened the pull request:
This implements suggestions from #18018 (see there for some discussion):

1. Clarifies the documentation about the learning_rate (different from all other implementations by a factor of 4).
2. Scales PCA initialization to have the same std as the random initialization. (Update: only issues future warning for now. Would change in v1.2.)
3. Issues future warning that PCA initialization will become default in v1.2.
4. Implements learning_rate='auto' that scales the learning rate with the sample size.
5. Issues future warning that learning_rate='auto' will become default in v1.2.

I would still have to implement unit tests for future warnings (haven't done it before) and add the changes to `whats_new` (not quite sure which of the above changes need to be mentioned there). But I'd like to get some feedback from the core developers about whether these suggested changes are all fine. @TomDLT @ogrisel

Update: tests added, changes added.

dkobak committed:
Rescale PCA initialization to std=1e-4

dkobak committed:
Add future warning about PCA init

dkobak committed:
Add learning_rate='auto' and comment on factor 4

dkobak committed:
Add future warning for the learning rate

dkobak committed:
Add whitespace

dkobak committed:
Remove whitespaces

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -681,12 +688,30 @@ def __init__(self, n_components=2, *, perplexity=30.0,
+            # See issue #18018
+            warnings.warn("The default initialization will change from "
+                          "random to PCA in 1.01.", FutureWarning)
+            self.init = 'random'
```
We avoid changing attributes that were defined in the init, to avoid unexpected effects when using hyperparameter search. So we prefer defining another variable:
```py
if self.init == 'warn':
    init = "random"
else:
    init = self.init
```

dkobak replied:
Wait, but it is not what the example here https://scikit-learn.org/stable/developers/contributing.html#change-the-default-value-of-a-parameter shows!

TomDLT replied:
The example does not overwrite `self.n_clusters`, it defines a new private attribute `self._n_clusters`.

dkobak replied:
Oooh! Indeed. Thanks, I entirely missed it. Pushed a fix now.

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -823,6 +848,9 @@ def _fit(self, X, skip_num_points=0):
             X_embedded = pca.fit_transform(X).astype(np.float32, copy=False)
+            # PCA is rescaled so that PC1 has standard deviation 1e-4 which is
+            # the default value for random initialization. See issue #18018.
+            X_embedded = X_embedded / np.std(X_embedded[:, 0]) * 1e-4
```
I am not sure if we can consider this change a bugfix, so we might also need a proper future warning cycle here.

dkobak replied:
You are right -- this is not a bugfix. All right, I made this a future warning. Take a look if it's thrown in the correct place.

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -656,9 +663,9 @@ class TSNE(BaseEstimator):
+                 early_exaggeration=12.0, learning_rate="warn", n_iter=1000,
                  n_iter_without_progress=300, min_grad_norm=1e-7,
-                 metric="euclidean", init="random", verbose=0,
+                 metric="euclidean", init="warn", verbose=0,
```
You need to update the docstring for `init` as well.

dkobak replied:
Update how? You mean the docstring should mention that the default is going to change in version 1.2? Otherwise the current docstring is fine.

TomDLT replied:
Yes, I would mention the future default.

dkobak replied:
OK -- added.

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -681,12 +688,30 @@ def __init__(self, n_components=2, *, perplexity=30.0,
+        if self.learning_rate == 'auto':
+            # See issue #18018
+            self.learning_rate = X.shape[0] / self.early_exaggeration / 4
+            self.learning_rate = np.max(self.learning_rate, 50)
```
Is 50 better than the previous value of 200 ?

dkobak replied:
I haven't extensively tested it (note that this only matters for relatively small sample sizes, otherwise N/48 heuristic will kick in). 50 corresponds to the default in other implementations including the "official" Barnes Hut implementation of Maaten (where it's 200 but without the factor of 4). Which is why openTSNE uses `max(50, N/12)` rule. But we can ask well keep 200 here for better consistency with the current sklearn default. I have no preference.

TomDLT replied:
Ok makes sense, 50 is fine.

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -681,12 +688,30 @@ def __init__(self, n_components=2, *, perplexity=30.0,
 
+        if self.init == 'warn':
+            # See issue #18018
+            warnings.warn("The default initialization will change from "
+                          "random to PCA in 1.01.", FutureWarning)
```
We will name versions 1.0, 1.1, 1.2, ... We also wait two releases to change behaviors.
```suggestion
            warnings.warn("The default initialization in TSNE will change "
                          "from 'random' to 'pca' in 1.2.", FutureWarning)
```

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -681,12 +688,30 @@ def __init__(self, n_components=2, *, perplexity=30.0,
+            self.init = 'random'
+        if self.learning_rate == 'warn':
+            # See issue #18018
+            warnings.warn("The default learning rate will change from "
+                          "200.0 to 'auto' in 1.01.", FutureWarning)
```
```suggestion
            warnings.warn("The default learning rate in TSNE will change "
                          "from 200.0 to 'auto' in 1.2.", FutureWarning)
```

TomDLT reviewed (commented):
Thanks for the pull-request!

dkobak committed:
Update sklearn/manifold/_t_sne.py

Co-authored-by: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

dkobak committed:
Update sklearn/manifold/_t_sne.py

Co-authored-by: Tom Dupré la Tour <tom.dupre-la-tour@m4x.org>

dkobak committed:
Future warning for PCA init

dkobak committed:
Do not overwrite default attributes

dkobak committed:
Remove trailing whitespaces

dkobak committed:
Ignore PCA/init future warnings in tests

dkobak committed:
Fix whitespaces before inline comments

dkobak committed:
Fix more whitespaces

dkobak committed:
Add tests for new future warnings and for lr=auto

dkobak committed:
Fix a bug in one test

dkobak committed:
Fix another bug

dkobak committed:
Split one test into two

dkobak committed:
Rename new tests

dkobak committed:
Correctly handle None to invoke default

dkobak committed:
Fix the auto learning rate text

dkobak committed:
Fix random state for assert_allclose

dkobak committed:
Fix max bug

dkobak committed:
Fix max bug

dkobak commented:
@TomDLT I fixed and added t-SNE tests and they seem to be working fine, but something is failing in `doctest`:

```
================================== FAILURES ===================================
____________________ [doctest] sklearn.manifold._t_sne.TSNE ____________________
...
UNEXPECTED EXCEPTION: FutureWarning("The default initialization in TSNE will change from 'random' to 'pca' in 1.2.")
```
and I don't know where to fix it :-/

dkobak committed:
Ignore future warnings in test

dkobak committed:
Ignore future warnings in more tests

dkobak committed:
Import ignore_warnings

dkobak committed:
Avoid future warnings in docstring test

dkobak commented:
Update: I fixed almost all testing errors, apart from the `Check Changelog` which I don't know how to fix and one more error under `Linux pylatest_pip_openblas_pandas` which I don't really understand. Would very much appreciate any advice from core developers @TomDLT @ogrisel ! Thanks.

cmarmo commented:
@dkobak the changelog check is failing because there is no entry in the [what's new file](https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/v1.0.rst) referring to this pull request. Thanks for your work!

dkobak commented:
@cmarmo thanks, I see! However, in this case most changes will only happen in 1.2. Should I put it into a what's new file for v 1.2? Should I put it into 1.0 into some dedicated "future changes" section? How does it work? The only thing that will change directly in 1.0 is that `learning_rate='auto'` will be implemented. Should I only mention that?

TomDLT commented:
You can describe all your changes in the v1.0 changelog, under the `manifold` section.

The other error is from a FutureWarning in a doctest. You can reproduce it with `pytest -Werror::FutureWarning  sklearn/manifold/`. You need to update the docstring example to avoid warnings.

dkobak committed:
Describe t-SNE changes in whats_new

dkobak committed:
Update the docstring example to avoid warnings

dkobak committed:
Fix newline for doctest

dkobak commented:
Thanks! All checks have passed!

TomDLT commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -954,7 +985,64 @@ def test_tsne_square_distances_futurewarning(metric, square_distances):
+        assert not record
+
+
+@pytest.mark.parametrize('sample_size', [500, 10000])
```
This test is too long, we have to keep the testing suite as short as possible. If a faster test is not possible, I would remove this test.
> [============================= slowest 20 durations =============================](https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=26885&view=logs&j=13650d7c-49e8-54f3-0598-c3480d1c1e4f&t=97f2162d-a0c4-54dc-9fce-2fceef7172e3&l=353)
**226.91s call     manifold/tests/test_t_sne.py::test_auto_learning_rate[10000]**
14.74s call     tests/test_common.py::test_estimators[SequentialFeatureSelector(estimator=LogisticRegression(C=1))-check_estimator_sparse_data]
13.22s call     ensemble/tests/test_voting.py::test_gridsearch
10.09s call     svm/tests/test_svm.py::test_svc_ovr_tie_breaking[NuSVC]
9.17s call     utils/tests/test_estimator_checks.py::test_check_estimator_clones
**7.92s call     manifold/tests/test_t_sne.py::test_auto_learning_rate[500]**
5.31s call     ensemble/tests/test_common.py::test_ensemble_heterogeneous_estimators_behavior[stacking-classifier]
5.25s call     ensemble/tests/test_bagging.py::test_classification
...

dkobak replied:
I see. I don't know how to make it faster so I removed the test.

TomDLT reviewed (commented):
Good job fixing the tests !

dkobak committed:
Delete the slow test

TomDLT commented on sklearn/manifold/_t_sne.py:
```diff
@@ -681,12 +691,35 @@ def __init__(self, n_components=2, *, perplexity=30.0,
+            self._learning_rate = np.maximum(self._learning_rate, 50)
+        else:
+            if not (self._learning_rate > 0):
+                raise ValueError("'learning_rate' must be a positive number "
```
Codecov is noting that this line is not covered by the tests.
Could you add a small test, testing that this error is raised when giving a non-positive learning rate?

TomDLT reviewed (approved):
This PR is looking good!

dkobak committed:
Add a test for negative learning rate

dkobak committed:
Add whitespace

dkobak committed:
Remove unused variable

TomDLT commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -1028,6 +1028,20 @@ def test_tsne_learning_rate_futurewarning(learning_rate):
+    if_value_error = False
+    try:
+        X_tr = TSNE(learning_rate=learning_rate).fit_transform(X)
+    except ValueError:
```
Thanks for the test. A more standard way to check the error is to use `pytest.raises`, see for example `test_distance_not_available`.

dkobak committed:
Use pytest.raises

dkobak committed:
Remove obsolete param

dkobak commented:
I added the test. Unfortunately, something else is failing now; so far I could not understand why adding this test resulted in some other tests suddenly failing :-(

dkobak commented:
Looks like it's `check_n_features_in_after_fitting` in `sklearn/utils/estimator_checks.py` running into the future warning. Why this only started to happen now is a mystery to me. Can I add `@ignore_warnings(category=FutureWarning)` there?

dkobak commented:
Oh I see, I think it's because there was another PR earlier today that allowed `check_n_features_in_after_fitting` to run on `manifold` classes... Arrrgh, now I am stuck because my branch does not have those recent pulls and I am not proficient enough in git to fix it :-/

TomDLT commented:
You can merge the main branch into your branch, using:
```bash
git fetch upstream  # assuming the scikit-learn remote is called upstream
git checkout tsne-defaults
git merge upstream/main
```

dkobak committed:
Merge remote-tracking branch 'upstream/main' into tsne-defaults

dkobak committed:
Ignore future warnings in check_n_features_in_after_fitting

dkobak commented:
Thanks for the tip! I updated my branch to the current main and added `@ignore_warnings(category=FutureWarning)` there. However, I noticed that the current main branch seems to fail some tests, so I don't know if my PR is going to have some failed checks because of that.

dkobak commented:
Update: it worked fine :)

TomDLT commented:
Remark: the proposed future default `init='pca'` is not compatible with `metric='precomputed'`, therefore:
- we might reconsider changing the default there. No strong opinion there.
- if we still change the default, we need to update the tests that use `metric='precomputed'` to not use the default init.

dkobak commented:
Thanks. This is a good point. I did not think of the `metric='precomputed'`. I suggest we make PCA init default unless the metric=precomputed. In the latter case, if `init` is not provided, it will default to random init. What do you think?

dkobak commented:
Update: hmm, maybe I did not think it through. Would this require something like default `init='auto'` that would then become `pca` or `random` depending on whether the metric is `precomputed`? This sounds more complicated than it should be...

Or can we simply make `init='pca'` default (in the future) but if `metric='precomputed'`, then instead of this error:
```
raise ValueError("The parameter init=\"pca\" cannot be "
                                 "used with metric=\"precomputed\".")
```
it prints a warning that `random` is going to be used instead?

dkobak commented:
Okay, apologies for spamming but this:
> if we still change the default, we need to update the tests that use metric='precomputed' to not use the default init.
makes a lot of sense and is what we should do IMHO. I can put `init='random'` in all those tests.

This only has the downside from the API point of view, because `TSNE(metric='precomputed')` will yield an error message. That's why I suggested changing the error into a warning and switching to `init='random'` in this case.

dkobak committed:
Use random init for precomputed metric in tests

dkobak committed:
Merge branch 'tsne-defaults' of https://github.com/dkobak/scikit-learn into tsne-defaults

dkobak committed:
Remove double arg

dkobak committed:
Fix one test

dkobak committed:
Init can be ndarray, so compare properly

dkobak commented:
Update: I added `init='random'` to those tests.

> This only has the downside from the API point of view, because TSNE(metric='precomputed') will yield an error message. That's why I suggested changing the error into a warning and switching to init='random' in this case.

I haven't implemented this yet -- will wait for the feedback. We can also keep at as.

dkobak commented:
@TomDLT Gentle ping -- when you have a moment, let me know what you think about this last change I did.

TomDLT commented:
I was initially not convinced by the `init='auto'`, but I realize `init='pca'` fails not only with precomputed neighbor graphs, but also with sparse arrays, so I am reconsidering it as a reasonable option.
Basically, `init='auto'` would:
- use `'pca'` if the input array X is dense
- use `'random'` if the input array X is sparse, or if `metric='precomputed'`

I prefer adding an explicit 'auto' option, rather than magically changing the behavior of 'pca' to 'random' with just a warning.

dkobak commented:
Good point about sparse inputs, I did not think of this.

Still, I am unsure about `init='auto'`. That sklearn's PCA cannot process  sparse input is the problem of the PCA estimator, not of t-SNE. (There are pull requests stalled for over a year that would add this functionality, by the way.) For precomputed distance matrices, one could in principle use PCoA (aka classical MDS) which is basically PCA on the distance matrix (assuming Euclidean distance). But I think this is not implemented in sklearn either. So if these two things were available in the PCA estimator, then `init='pca'` in t-SNE would make sense for all possible inputs...

To be honest, my preferred solution right now is to do nothing :-) The error messages are informative, so anybody running t-SNE on a sparse input or with a precomputed metric would see that the initialization fails and they need to change initialization to `init='random'`. 

But if you think `init='auto'` is a better way, I can implement that.

TomDLT commented:
Good point. If we change the default to 'pca', I would suggest to add some error messages more specific to TSNE, suggesting to use 'random' initialization instead.

Let's see what other developers prefer here.

dkobak committed:
error message on sparse input with pca init

dkobak committed:
linting fix

dkobak committed:
error type fix

dkobak committed:
future warning fix

dkobak commented:
OK. For precomputed input with PCA init there was already an error message before. I have now added an error message (+test) for PCA init with sparse input.

Will be waiting for other opinions/approvals @ogrisel @amueller. Cheers.

dkobak commented:
Oh, there is something I forgot to mention in the original issue: after implementing the `learning_rate = n/12` heuristic in openTSNE and FIt-SNE we realized that 750 iterations is enough for all practical purposes and made `iter=750` the default over there in both implementations (see a LONG discussion here https://github.com/KlugerLab/FIt-SNE/issues/88).

So we could also adopt the same convention here, cutting down the number of default iterations from 1000 to 750. This of course would need to go through a deprecation cycle, together with the `learning_rate='auto'`. What do you think?

In case you think it's a good idea, I am wondering if the deprecation cycle needs to be implemented via `learning_rate='warn'`. Given that this is tied to the `learning_rate` change, can the learning rate FutureWarning mention that the max_iter will change to 750 together with the future learning rate change? Without an additional learning rate warning?

dkobak commented:
@TomDLT Hi Tom, let me know what you think about the `n_iter` (see comment above) when you have a moment.

Also, whom else could I maybe ping here for review?

dkobak commented:
I noticed that @jnothman has been participating in other t-SNE-related PRs, so I thought I would ping him here as well. Thanks!

dkobak commented:
Dear all, is there anything I can do to help bring this PR forward?

TomDLT commented:
> Before merging, we generally require two core developers to agree that your pull request is desirable and ready. [Please be patient](http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention), as we mostly rely on volunteered time from busy core developers. (You are also welcome to help us out with [reviewing other PRs](http://scikit-learn.org/dev/developers/contributing.html#code-review-guidelines).)

You can try to ping other developers for reviews, but there is unfortunately no guarantee that they will have time for this PR. Sometimes, splitting large PR into multiple small independent PRs can simplify the review process, but I am not sure it makes sense for this one. Another possibility is to raise interest by linking this PR in other related discussions.

thomasjpfan commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -249,6 +249,7 @@ def test_trustworthiness():
 
 @pytest.mark.parametrize("method", ['exact', 'barnes_hut'])
 @pytest.mark.parametrize("init", ('random', 'pca'))
+@ignore_warnings(category=FutureWarning)  # Delete in 1.2
```
I think it's better to be more explicit here. We do not want to catch `FutureWarnings` that are not related to the deprecation.

```suggestion
# TODO: Remove filterwarning in 1.2
@pytest.mark.filterwarnings("ignore:.*TSNE will change from.*:FutureWarning")
@pytest.mark.parametrize("method", ['exact', 'barnes_hut'])
@pytest.mark.parametrize("init", ('random', 'pca'))
```

thomasjpfan reviewed (commented):
This PR is deprecating three things: two defaults value and the PCA SD change.  I suggest removing the PCA SD change for now and keep the deprecation for `learning_rate` and `init`. In the user guide (`doc/modules/manifold.rst`), we need to describe `learning_rate='auto'` with references.

In a future PR we can update the user guide for the PCA SD change and add the warning + tests.

dkobak committed:
Update sklearn/manifold/_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Add TODO comments and docstrings for tests

dkobak committed:
Merge remote-tracking branch 'upstream/main' into tsne-defaults

dkobak commented:
@thomasjpfan Thanks a lot for reviewing! I pushed your suggestions and added TODO comments and docstrings as you suggested everywhere else too.

> This PR is deprecating three things: two defaults value and the PCA SD change. I suggest removing the PCA SD change for now and keep the deprecation for learning_rate and init. In the user guide (doc/modules/manifold.rst), we need to describe learning_rate='auto' with references. In a future PR we can update the user guide for the PCA SD change and add the warning + tests.

Sorry, I am not sure I understand the rationale here. I already have everything regarding PCA SD implemented here, what's the point of taking it out? Also, I would be uncomfortable setting PCA init as default if it does not get scaled to correct SD. To be honest, I think these changes should go together.

> In the user guide (doc/modules/manifold.rst), we need to describe learning_rate='auto' with references.

This I can do.

Also, what do you think about this suggested change (not yet implemented in this PR):

> Oh, there is something I forgot to mention in the original issue: after implementing the learning_rate = n/12 heuristic in openTSNE and FIt-SNE we realized that 750 iterations is enough for all practical purposes and made n_iter=750 the default over there in both implementations (see a LONG discussion here KlugerLab/FIt-SNE#88).
> 
> So we could also adopt the same convention here, cutting down the number of default iterations from 1000 to 750. This of course would need to go through a deprecation cycle, together with the learning_rate='auto'. What do you think?
>
> In case you think it's a good idea, I am wondering if the deprecation cycle needs to be implemented via n_iter='warn'. Given that this is tied to the learning_rate change, can the learning rate FutureWarning mention that the n_iter will change to 750 together with the future learning rate change? Without an additional n_iter future warning?

dkobak committed:
Describe learning_rate=auto in t-SNE doc

dkobak committed:
Fix ignoring future warning

dkobak commented:
@thomasjpfan Could you clarify what you meant by "removing the PCA SD change for now"? See also my comment above for more considerations. Everything is fixed btw. Cheers!

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -517,13 +517,21 @@ class TSNE(BaseEstimator):
+        ours. So our learning_rate=200 corresponds to learning_rate=800 in
+        those other implementations.
+        The 'auto' option sets the learning_rate to N / early_exaggeration / 4,
+        where N is the sample size, following Belkina et al. 2019 and
+        Kobak et al. 2019, Nature Communications (or to 50.0, if
+        N / early_exaggeration / 4 < 50). This will become default in 1.2.
```
Can we move this references into the References section below and link it here?

https://github.com/scikit-learn/scikit-learn/blob/962bd9a401bfee1f2d8e7e832018a75424b5bbe2/sklearn/manifold/_t_sne.py#L638

dkobak replied:
Makes sense, fixed.

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -517,13 +517,21 @@ class TSNE(BaseEstimator):
+        etc.) use a definition of learning_rate that is 4 times smaller than
+        ours. So our learning_rate=200 corresponds to learning_rate=800 in
+        those other implementations.
+        The 'auto' option sets the learning_rate to N / early_exaggeration / 4,
```
```suggestion
        The 'auto' option sets the learning_rate to
        `max(N / early_exaggeration / 4, 50)`. This will become default
        in 1.2.
```

thomasjpfan commented on sklearn/manifold/_t_sne.py:
```diff
@@ -817,13 +854,20 @@ def _fit(self, X, skip_num_points=0):
                       random_state=random_state)
             X_embedded = pca.fit_transform(X).astype(np.float32, copy=False)
-        elif self.init == 'random':
+            # PCA is rescaled so that PC1 has standard deviation 1e-4 which is
```
```suggestion
			# TODO: Update in 1.2
            # PCA is rescaled so that PC1 has standard deviation 1e-4 which is
```

thomasjpfan commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -261,6 +263,8 @@ def test_preserve_trustworthiness_approximately(method, init):
 
 
+# TODO: Remove ignore_warnings in 1.2 when default init changes to 'pca'
+@ignore_warnings(category=FutureWarning)
```
Can we update the `ignore_warnings` here to use `pytest.mark.filterwarnings` so we are explicit about which warnings we are catching?

```suggestion
@pytest.mark.filterwarnings("ignore:.*TSNE will change.*:FutureWarning")
```

thomasjpfan commented on sklearn/manifold/tests/test_t_sne.py:
```diff
@@ -275,7 +279,9 @@ def test_optimization_minimizes_kl_divergence():
     assert kl_divergences[2] <= kl_divergences[1]
 
 
+# TODO: Remove ignore_warnings in 1.2 when default init changes to 'pca'
```
With the explicit filterwarning, we can keep the comment simpler:

```suggestion
# TODO: Remove filterwarnings in 1.2
@pytest.mark.filterwarnings("ignore:.*TSNE will change.*:FutureWarning")
```

thomasjpfan reviewed (commented):
> Could you clarify what you meant by "removing the PCA SD change for now"? See also my comment above for more considerations. Everything is fixed btw. Cheers!

What I meant was that the commented out PCA change can be its own pull request. But looking at this again, I am okay with leaving it in.

I can see that the depredations in this PR are all related, but they could be done with three separate PRs. This makes it easier to review and helps with merging faster. In general, a PR with bigger scope increases the chances of something blocking it from merging.

> So we could also adopt the same convention here, cutting down the number of default iterations from 1000 to 750. This of course would need to go through a deprecation cycle, together with the learning_rate='auto'. What do you think?

We can work on this in a follow up PR. This PR is already a net improvement as is.

As for the review, I left comments about using `pytest.mark.filterwarnings` instead of `ignore_warnings` that applies to all the tests. Otherwise this looks good to go.

dkobak committed:
Update sklearn/manifold/_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
Update sklearn/manifold/tests/test_t_sne.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

dkobak committed:
linting fix

dkobak committed:
format references

dkobak committed:
linting fix

dkobak committed:
replace ignore_warnings with filterwarning

dkobak committed:
handle future warnings in tests in other files

dkobak committed:
import pytest

dkobak committed:
roll back because of failing test

dkobak commented:
@thomasjpfan Thanks a lot! I changed the handling of the future warnings. All checks pass.

> We can work on this in a follow up PR. This PR is already a net improvement as is.

My suggestion is simply to replace
```
        if self.learning_rate == 'warn':
            # See issue #18018
            warnings.warn("The default learning rate in TSNE will change "
                          "from 200.0 to 'auto' in 1.2.", FutureWarning)
```
with
```
        if self.learning_rate == 'warn':
            # TODO: Change the n_iter to 750 in 1.2.
            # See issue #18018
            warnings.warn("The default learning rate in TSNE will change "
                          "from 200.0 to 'auto' in 1.2. At the same time, the default 
                          "number of iterations will decrease from 1000 to 750",
                          FutureWarning)
```
I think this does not deserve its own PR... There would nothing else to do really at this point. Or what do you think?

TomDLT commented:
I am not fully convinced that changing the default number of iteration from 1000 to 750 is necessary. It would probably benefit from a dedicated discussion in a small separate PR.

dkobak commented:
> I am not fully convinced that changing the default number of iteration from 1000 to 750 is necessary. It would probably benefit from a dedicated discussion in a small separate PR.

Fair enough. It won't have any other consequences apart from increasing the runtime by 25%. I'm fine merging this PR without this change if you guys prefer that.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/19491

thomasjpfan committed:
CLN Remove comment for check_n_features_in_after_fitting

thomasjpfan reviewed (approved):
Thank you for working on this issue @dkobak !

LGTM
